File: Lab4/lab4.2/TarsosDSP/android/src/main/java/be/tarsos/dsp/io/android/AudioDispatcherFactory.java
package be.tarsos.dsp.io.android;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;

import be.tarsos.dsp.AudioDispatcher;
import be.tarsos.dsp.io.TarsosDSPAudioFormat;

public class AudioDispatcherFactory {

    public static AudioDispatcher fromDefaultMicrophone(final int sampleRate, final int audioBufferSize, final int bufferOverlap) {
        int minBufferSize = AudioRecord.getMinBufferSize(sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT);
        int bufferSize = Math.max(minBufferSize, audioBufferSize);
        
        AudioRecord audioRecord = new AudioRecord(MediaRecorder.AudioSource.MIC, sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, bufferSize);
        
        TarsosDSPAudioFormat format = new TarsosDSPAudioFormat(sampleRate, 16, 1, true, false);
        
        AndroidAudioInputStream audioStream = new AndroidAudioInputStream(audioRecord, format);
        
        return new AudioDispatcher(audioStream, audioBufferSize, bufferOverlap);
    }
}

File: Lab5/android-app/app/src/main/java/com/example/lab5/AudioDispatcherFactory.kt
package com.example.lab5

import android.annotation.SuppressLint
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import be.tarsos.dsp.AudioDispatcher
import be.tarsos.dsp.io.TarsosDSPAudioFormat
import be.tarsos.dsp.io.UniversalAudioInputStream
import java.io.InputStream

object AudioDispatcherFactory {
    @SuppressLint("MissingPermission")
    fun fromDefaultMicrophone(sampleRate: Int, bufferSize: Int, overlap: Int): AudioDispatcher {
        val minBufferSize = AudioRecord.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        )
        val actualBufferSize = minBufferSize.coerceAtLeast(bufferSize * 2)
        val audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
            actualBufferSize
        )
        audioRecord.startRecording()

        val inputStream = object : InputStream() {
            private val buf = ByteArray(actualBufferSize)
            override fun read(): Int {
                val b = ByteArray(1)
                val r = read(b, 0, 1)
                return if (r == -1) -1 else b[0].toInt() and 0xFF
            }
            override fun read(b: ByteArray, off: Int, len: Int): Int {
                return audioRecord.read(b, off, len)
            }
            override fun close() {
                audioRecord.stop()
                audioRecord.release()
            }
        }

        val format = TarsosDSPAudioFormat(sampleRate.toFloat(), 16, 1, true, false)
        val audioStream = UniversalAudioInputStream(inputStream, format)
        return AudioDispatcher(audioStream, bufferSize, overlap)
    }
}